src/topojson/topojson.ml

```diff
@@ -698,24 +696,23 @@ module Make (J : Intf.Json) = struct
     let junction_table : (Position.t, Set.t) Hashtbl.t =
       Hashtbl.create (List.length lines)
     in
-    let junction : Position.t list = [] in
+    let junction : Position.t list ref = ref [] in
     List.iter
       (fun line ->
         let line_length = Array.length line in
-        for i = 1 to line_length - 2 do
+        for i = 0 to line_length - 1 do
           let point1 = line.(i) in
-          let point2 = line.(i - 1) in
-          let point3 = line.(i + 1) in
+          let point2 = if i > 0 then line.(i - 1) else line.(line_length - 1) in
+          let point3 = if i < line_length - 1 then line.(i + 1) else line.(0) in
           if Hashtbl.mem junction_table point1 then
             let neighbors = Hashtbl.find junction_table point1 in
             Hashtbl.replace junction_table point1
```

Comment:
Instead of adding the neighbours to an ever growing set, I think we want to keep them distinct so we know what two neighbours belong together. Consider the problem of having two lines `ABC` and `CBA`. `B` in this case is _not_ a junction because the neighbours `A` and `C` are the same in both lines.